File: app/src/main/java/social/entourage/android/api/ApiError.java
package social.entourage.android.api;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import retrofit2.Response;

import java.io.IOException;

public class ApiError {
    public final String code;

    private ApiError(String code) {
        this.code = code;
    }

    public static ApiError fromResponse(Response response) {
        ApiError error;

        try {
            error = new Gson().fromJson(response.errorBody().string(), ApiErrorContainer.class).error;
        } catch (IOException | JsonSyntaxException | NullPointerException e) {
            error = null;
        }

        if (error == null || error.code == null) {
            error = new ApiError("OTHER");
        }

        return error;
    }

    private static class ApiErrorContainer {
        private ApiError error;
    }
}

File: app/src/main/java/social/entourage/android/api/ApiError.kt
package social.entourage.android.api

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import retrofit2.Response
import java.io.IOException

class ApiError private constructor(code: String) {
    val code: String?

    private class ApiErrorContainer {
        val error: ApiError? = null
    }

    companion object {
        fun fromResponse(response: Response<*>): ApiError {
            response.errorBody()?.string()?.let {
                try {
                    val error: ApiError? = Gson().fromJson(it, ApiErrorContainer::class.java).error
                    if (error?.code != null) {
                        return error
                    }
                } catch (e: IOException) {
                } catch (e: JsonSyntaxException) {
                } catch (e: NullPointerException) {
                }
            }
            return ApiError("OTHER")
        }
    }

    init {
        this.code = code
    }
}